src/kernel/arch/x86_64/cpu/tss.c

```diff
@@ -0,0 +1,65 @@
+#include <logging.h>
+#include <tss.h>
+
+extern uint64_t gdt64[];
+extern uint64_t stack;
+
+tss_t kernel_tss;
+
+void initialize_tss(){
+
+    loglinef(Verbose, "(%s) Initializing tss", __FUNCTION__);
+
+    // These fields are reserved and must be set to 0
+    kernel_tss.reserved0 = 0x00;
+    kernel_tss.reserved1 = 0x00;
+    kernel_tss.reserved2 = 0x00;
+    kernel_tss.reserved3 = 0x00;
+    kernel_tss.reserved4 = 0x00;
+
+    // The rspX are used when there is a privilege change from a lower to a higher privilege
+    // Rsp contain the stack for that privilege level.
+    // We use only privilege level 0 and 3, so rsp1 and rsp2 can be left as 0
+    // Every thread will have it's own rsp0 pointer
+    kernel_tss.rsp0 = 0x0;
+    kernel_tss.rsp1 = 0x0;
+    kernel_tss.rsp2 = 0x0;
+    // istX are the Interrup stack table,  unless some specific cases they can be left as 0
+    // See intel manual chapter 5
+    kernel_tss.ist1 = 0x0;
+    kernel_tss.ist2 = 0x0;
+    kernel_tss.ist3 = 0x0;
+    kernel_tss.ist4 = 0x0;
+    kernel_tss.ist5 = 0x0;
+    kernel_tss.ist6 = 0x0;
+    kernel_tss.ist7 = 0x0;
+    // Can be left as 0 for now
+    kernel_tss.io_bitmap_offset = 0x0;
+}
+
+void load_tss() {
+    // Fields explanation (each entry is 64bit)
+    // TSS_ENTRY_LOW:
+    // 0:15 -> Limit (first 15 bits) should be 0xFFFF
+    // 16:39 -> First 24 bits of kernel_tss address
+    // 40:47 -> Type 4 bits in our case is 1001, 0,  DPL should be 0 , P = 1
+    // 48:55 -> Limit (last 4 bits) can be 0, AVL=available to OS we leave it as 0, 53:54 are 0, 55 G (Granularity)
+    // 55:63 -> Bits 25:31 of the kernel_tss base address
+    // TSS_ENTRY_HIGH
+    // 0:31 -> kernel_tss bits 32:63
+    // 32:63 -> Reserved / 0
+    // TYPE: 1001 (64Bit TSS Available)
+    // BASE_ADDRESS: kernel_tss
+    // LIMIT 16:19 0 DPL: 0 P: 1 G:0
+    gdt64[TSS_ENTRY_LOW] = 0x00;
+    gdt64[TSS_ENTRY_HIGH] = 0x00;
+    gdt64[TSS_ENTRY_LOW] = ((((uint64_t) &kernel_tss>>24) & 0xFF)<<56)  /*bits 56:63*/ | ((uint64_t) (0 & 0xFF) << 48) /* bits 48:55 */ | ((uint64_t)0x89 << 40) /* bits 40:47*/|  ((((uint64_t) &kernel_tss) & 0xFFFFFF) << 16) /* bits 15:39 */ | (uint64_t) 0x70 /* Base */;
```

Comment:
this looks like lisp! hahaha
It might be easier to understand if you break this down into a few smaller steps (i.e. local variables) and then assemble those.

Reply:
ok, but then shouldn't be better to create a struct with bitfields? 
If i start to create local variable is more or less the same aestethically. 
Am i wrong? 

Btw yeah probably at least using variables, shoul make it more readable!